lib/transform.js

```diff
@@ -87,235 +42,47 @@ function transform(source, template, options = {}) {
-    let handlerMethod = properties.filter(path => isMethod(path, methodName))[0];
-    if (handlerMethod) {
-      throw new SilentError(`Using \`${methodName}()\` is not supported in tagless components`);
-    }
-  }
-
-  // analyze `elementId`, `attributeBindings`, `classNames` and `classNameBindings`
-  let elementId = findElementId(properties);
-  debug('elementId: %o', elementId);
-
-  let attributeBindings = findAttributeBindings(properties);
-  debug('attributeBindings: %o', attributeBindings);
-
-  let classNames = findClassNames(properties);
-  debug('classNames: %o', classNames);
-
-  let classNameBindings = findClassNameBindings(properties);
-  debug('classNameBindings: %o', classNameBindings);
-
-  let templateAST = templateRecast.parse(template);
-
-  // set `tagName: ''`
-  let tagNamePath = j(properties)
-    .find(j.ObjectProperty)
-    .filter(path => path.parentPath === properties)
-    .filter(path => isProperty(path, 'tagName'));
-
-  if (tagNamePath.length === 1) {
-    j(tagNamePath.get('value')).replaceWith(j.stringLiteral(''));
-  } else {
-    properties.unshift(j.objectProperty(j.identifier('tagName'), j.stringLiteral('')));
-  }
-
-  // remove `elementId`, `attributeBindings`, `classNames` and `classNameBindings`
-  j(properties)
-    .find(j.ObjectProperty)
-    .filter(path => path.parentPath === properties)
-    .filter(
-      path =>
-        isProperty(path, 'elementId') ||
-        isProperty(path, 'attributeBindings') ||
-        isProperty(path, 'classNames') ||
-        isProperty(path, 'classNameBindings')
-    )
-    .remove();
-
-  let newSource = root.toSource();
-
-  // wrap existing template with root element
-  let classNodes = [];
-  if (options.hasComponentCSS) {
-    classNodes.push(b.mustache('styleNamespace'));
-  }
-  for (let className of classNames) {
-    classNodes.push(b.text(className));
-  }
-  classNameBindings.forEach(([truthy, falsy], property) => {
-    if (!truthy) {
-      classNodes.push(b.mustache(`unless this.${property} "${falsy}"`));
-    } else {
-      classNodes.push(b.mustache(`if this.${property} "${truthy}"${falsy ? ` "${falsy}"` : ''}`));
-    }
-  });
-
-  let attrs = [];
-  if (elementId) {
-    attrs.push(b.attr('id', b.text(elementId)));
-  }
-  attributeBindings.forEach((value, key) => {
-    attrs.push(b.attr(key, b.mustache(`this.${value}`)));
-  });
-  if (classNodes.length === 1) {
-    attrs.push(b.attr('class', classNodes[0]));
-  } else if (classNodes.length !== 0) {
-    let parts = [];
-    classNodes.forEach((node, i) => {
-      if (i !== 0) parts.push(b.text(' '));
-      parts.push(node);
+  if (result) {
+    let {
+      newSource,
+      tagName,
+      elementId,
+      classNames,
+      classNameBindings,
+      attributeBindings,
+    } = result;
+    let newTemplate = transformTemplate({
+      template,
+      tagName,
+      elementId,
+      classNames,
+      classNameBindings,
+      attributeBindings,
```

Comment:
Note: I cannot use object spread here because that's unsupported in node 8. But I guess we can drop support in a separate PR?

Reply:
I've taken the liberty of simplifying it a little more, by moving the `template` and `option` arguments out of the named arguments block again. That way, we can directly pass in `result` without the need for object spread.

Reply:
btw I had dropped Node 8 support in https://github.com/ember-codemods/tagless-ember-components-codemod/pull/48 😉